File: codebase/Forest/Composition/ViewRegistry.cs
/**
 * Copyright 2014 vdimensions.net.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
using System;
using System.Collections.Concurrent;
using System.Diagnostics;
using System.Linq;

using Forest.Stubs;


namespace Forest.Composition
{
    internal sealed class ViewRegistry : _ViewRegistry
    {
        private struct ViewEntry : IViewEntry
        {
            [DebuggerBrowsable(DebuggerBrowsableState.Never)]
            private readonly string id;
            [DebuggerBrowsable(DebuggerBrowsableState.Never)]
            private readonly IContainer container;
            [DebuggerBrowsable(DebuggerBrowsableState.Never)]
            private readonly IViewDescriptor descriptor;

            public ViewEntry(string id, IContainer container, IViewDescriptor descriptor)
            {
                this.id = id;
                this.container = container;
                this.descriptor = descriptor;
            }

            public string ID { get { return id; } }
            public Type Type { get { return descriptor.ViewType; } }
            public Type ViewModelType { get { return descriptor.ViewModelType; } }
            public IContainer Container { get { return container; } }
            public IViewDescriptor Descriptor { get { return descriptor; } }
        }

        [DebuggerBrowsable(DebuggerBrowsableState.Never)]
        private readonly ILogger logger;

        private readonly ConcurrentDictionary<string, IViewEntry> registry = new ConcurrentDictionary<string, IViewEntry>(StringComparer.Ordinal);
        private readonly ConcurrentDictionary<Type, IViewEntry> reverseViewModelToViewMap = new ConcurrentDictionary<Type, IViewEntry>();
        private readonly IForestContext context;

        public ViewRegistry(IForestContext context)
        {
            this.context = context;
            this.logger = context.LoggerFactory.GetLogger<ViewRegistry>();
        }

        #region Implementation of _ViewRegistry
        public IViewEntry Lookup(string id)
        {
            IViewEntry result;
            return registry.TryGetValue(id, out result) ? result : null;
        }
        public IViewEntry Lookup(Type viewModelType)
        {
            IViewEntry result;
            if (!reverseViewModelToViewMap.TryGetValue(viewModelType, out result))
            {
                var candidatePairs = reverseViewModelToViewMap.Where(x => x.Key.IsAssignableFrom(viewModelType)).ToArray();
                if (candidatePairs.Length == 1)
                {
                     return candidatePairs[0].Value;
                }
            }
            return null;
        }

        public void Register(string id, Type viewType, IContainer container)
        {
            var entry = new ViewEntry(id, container, this.context.GetDescriptor(viewType));
            if (registry.TryAdd(id, entry))
            {
                reverseViewModelToViewMap.TryAdd(entry.ViewModelType, entry);
                //logger.Warn("A view with id '{0}' has already been configured. Overwriting the existing data.", id);
            }
        }
        public void Register(IViewDescriptor descriptor, IContainer container)
        {
            var id = descriptor.ViewAttribute.ID;
            var entry = new ViewEntry(id, container, descriptor);
            if (registry.TryAdd(id, entry))
            {
                reverseViewModelToViewMap.TryAdd(entry.ViewModelType, entry);
                //logger.Warn("A view with id '{0}' has already been configured. Overwriting the existing data.", id);
            }
        }

        public void Unregister(string id)
        {
            IViewEntry entry;
            if (!registry.TryRemove(id, out entry))
            {
                logger.Warn("Attemting to remove a non-registered view id: \"{0}\"", id);
                IViewEntry idToRemove;
                reverseViewModelToViewMap.TryRemove(entry.ViewModelType, out idToRemove);
            }
        }
        #endregion
    }
}

File: codebase/core/main/src/Api/ViewRegistry.fs
//
// Copyright 2014-2019 vdimensions.net.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
namespace Forest

open System
open Forest.Collections


module ViewRegistry =
    let register<'T when 'T :> IView> (reg : IViewRegistry) =
        reg.Register<'T>()
    let registerViewType (viewType : Type) (reg : IViewRegistry) =
        reg.Register(viewType)

    let internal getDescriptorByName (name : vname) (reg : IViewRegistry) = 
        reg.GetDescriptor name

    let internal getDescriptorByType (viewType : Type) (reg : IViewRegistry) = 
        reg.GetDescriptor viewType

    let getDescriptor (viewHandle : ViewHandle) =
        match viewHandle with
        | ByName n -> getDescriptorByName n
        | ByType t -> getDescriptorByType t

    let internal resolve (descriptor : IViewDescriptor) (model : obj option) (reg : IViewRegistry) =
        match model with
        | None -> reg.Resolve descriptor
        | Some m -> reg.Resolve (descriptor, m)
